File: pcvr/Services/SongInstaller.cs
using HMUI;
using Newtonsoft.Json;
using SongCore;
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using UnityEngine;

namespace SnipeFeed.PC.Services
{
    internal sealed class SongInstallResult
    {
        public bool Success { get; set; }
        public string Error { get; set; } = "";
        public BeatmapLevel Level { get; set; }
    }

    internal static class SongInstaller
    {
        private const long MaxExtractedBytes = 512L * 1024 * 1024;
        private static readonly HttpClient Client = new HttpClient { Timeout = TimeSpan.FromSeconds(120) };

        static SongInstaller()
        {
            Client.DefaultRequestHeaders.UserAgent.ParseAdd("SnipeFeed-PC/2.0.0");
        }

        public static BeatmapLevel GetInstalledLevel(string hash)
        {
            if (string.IsNullOrWhiteSpace(hash)) return null;
            // Custom level IDs embed the hash uppercase; SongCore versions
            // differ on normalizing their argument, so try the uppercase
            // spelling first and the caller's spelling second.
            return Loader.GetLevelByHash(hash.ToUpperInvariant()) ?? Loader.GetLevelByHash(hash);
        }

        // The hash arrives from feed JSON; it becomes a URL segment and an
        // install folder name, so anything but exactly 40 hex chars is refused.
        private static bool IsValidHash(string hash)
        {
            if (string.IsNullOrEmpty(hash) || hash.Length != 40) return false;
            foreach (var c in hash)
            {
                var hex = (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
                if (!hex) return false;
            }
            return true;
        }

        public static async Task<SongInstallResult> DownloadAndInstallAsync(string hash)
        {
            hash = (hash ?? "").Trim().ToLowerInvariant();
            if (hash.Length == 0) return Fail("This score is not a downloadable custom song.");
            if (!IsValidHash(hash)) return Fail("This score has an invalid map hash.");

            var existing = GetInstalledLevel(hash);
            if (existing != null) return new SongInstallResult { Success = true, Level = existing };

            string temp = null;
            try
            {
                var metadataResponse = await Client.GetAsync("https://api.beatsaver.com/maps/hash/" + Uri.EscapeDataString(hash));
                if ((int)metadataResponse.StatusCode == 404) return Fail("Map not found on BeatSaver.");
                if (!metadataResponse.IsSuccessStatusCode) return Fail("BeatSaver lookup failed (HTTP " + (int)metadataResponse.StatusCode + ").");

                var map = JsonConvert.DeserializeObject<BeatSaverMap>(await metadataResponse.Content.ReadAsStringAsync());
                // Only the version matching the score's hash: installing a
                // different version would never match the post-install lookup
                // and silently changes the map under the score.
                var version = map?.versions?.FirstOrDefault(v => string.Equals(v.hash, hash, StringComparison.OrdinalIgnoreCase));
                if (version == null || string.IsNullOrWhiteSpace(version.downloadURL))
                    return Fail("This score's map version is no longer available on BeatSaver.");

                var zipBytes = await Client.GetByteArrayAsync(version.downloadURL);
                if (zipBytes == null || zipBytes.Length == 0) return Fail("The map download was empty.");

                // Unity APIs (Application.dataPath) before leaving the main
                // thread; the disk work below runs on the pool.
                var customLevels = Path.GetFullPath(Path.Combine(Application.dataPath, "CustomLevels"));
                var shortHash = hash.Substring(0, 10);
                var target = Path.Combine(customLevels, "SnipeFeed_" + shortHash);
                temp = Path.Combine(customLevels, ".snipefeed_tmp_" + shortHash);
                var tempForWork = temp;

                // Deleting, decompressing and writing a multi-MB map is
                // seconds of blocking IO — off the render thread, or every
                // install freezes VR. Extract to a temp sibling and move into
                // place so a failure never leaves a half-written level folder.
                await Task.Run(() =>
                {
                    Directory.CreateDirectory(customLevels);
                    if (Directory.Exists(tempForWork)) Directory.Delete(tempForWork, true);
                    Directory.CreateDirectory(tempForWork);
                    ExtractZipSafely(zipBytes, tempForWork);
                    if (Directory.Exists(target)) Directory.Delete(target, true);
                    // Windows can hold the deleted dir in a pending state
                    // briefly; a short retry absorbs that race.
                    for (var attempt = 0; ; attempt++)
                    {
                        try { Directory.Move(tempForWork, target); break; }
                        catch (IOException) when (attempt < 3) { Thread.Sleep(50); }
                    }
                });
                temp = null; // moved into place; nothing to clean up

                Plugin.Log?.Info("Installed map " + hash + " to " + target);
                await RefreshSongCore();

                return new SongInstallResult
                {
                    Success = true,
                    Level = GetInstalledLevel(hash)
                };
            }
            catch (TaskCanceledException)
            {
                CleanupPartialInstall(temp);
                return Fail("Map download timed out.");
            }
            catch (Exception ex)
            {
                Plugin.Log?.Error("Map install failed: " + ex);
                CleanupPartialInstall(temp);
                return Fail("Map install failed: " + ex.Message);
            }
        }

        private static void CleanupPartialInstall(string temp)
        {
            if (temp == null) return;
            try
            {
                if (Directory.Exists(temp)) Directory.Delete(temp, true);
            }
            catch (Exception ex)
            {
                Plugin.Log?.Warn("Couldn't remove a partial install folder: " + ex.Message);
            }
        }

        // Declared sizes in the central directory can be forged; the only
        // trustworthy count is bytes actually produced by decompression.
        private static void CopyBounded(Stream input, Stream output, ref long totalWritten)
        {
            var buffer = new byte[81920];
            int read;
            while ((read = input.Read(buffer, 0, buffer.Length)) > 0)
            {
                totalWritten += read;
                if (totalWritten > MaxExtractedBytes)
                    throw new InvalidDataException("Map archive is unreasonably large.");
                output.Write(buffer, 0, read);
            }
        }

        private static void ExtractZipSafely(byte[] bytes, string target)
        {
            var root = Path.GetFullPath(target).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar) + Path.DirectorySeparatorChar;
            using (var memory = new MemoryStream(bytes, false))
            using (var archive = new ZipArchive(memory, ZipArchiveMode.Read))
            {
                // A hostile or corrupt archive must not be able to fill the
                // disk; real beatmaps are tens of MB at most.
                long declaredTotal = 0;
                foreach (var entry in archive.Entries)
                {
                    declaredTotal += entry.Length;
                    if (declaredTotal > MaxExtractedBytes)
                        throw new InvalidDataException("Map archive is unreasonably large.");
                }

                long actualTotal = 0;
                foreach (var entry in archive.Entries)
                {
                    if (string.IsNullOrEmpty(entry.FullName)) continue;
                    var destination = Path.GetFullPath(Path.Combine(root, entry.FullName.Replace('/', Path.DirectorySeparatorChar)));
                    if (!destination.StartsWith(root, StringComparison.OrdinalIgnoreCase))
                        throw new InvalidDataException("Map archive attempted to write outside its install folder.");

                    if (entry.FullName.EndsWith("/", StringComparison.Ordinal) || entry.FullName.EndsWith("\\", StringComparison.Ordinal))
                    {
                        Directory.CreateDirectory(destination);
                        continue;
                    }

                    var parent = Path.GetDirectoryName(destination);
                    if (!string.IsNullOrEmpty(parent)) Directory.CreateDirectory(parent);
                    using (var input = entry.Open())
                    using (var output = new FileStream(destination, FileMode.Create, FileAccess.Write, FileShare.None))
                        CopyBounded(input, output, ref actualTotal);
                }
            }
        }

        private static async Task RefreshSongCore()
        {
            if (Loader.Instance == null) return;

            var completion = new TaskCompletionSource<bool>();
            Action<Loader, System.Collections.Concurrent.ConcurrentDictionary<string, BeatmapLevel>> handler = null;
            handler = (loader, levels) =>
            {
                Loader.SongsLoadedEvent -= handler;
                completion.TrySetResult(true);
            };

            Loader.SongsLoadedEvent += handler;
            Loader.Instance.RefreshSongs(false);

            var finished = await Task.WhenAny(completion.Task, Task.Delay(15000));
            Loader.SongsLoadedEvent -= handler;
            if (finished != completion.Task)
                Plugin.Log?.Warn("Timed out waiting for SongCore refresh; the map may appear a little later.");
        }

        // Mirrors the Quest mod's Installer (itself modeled on BeatLeader's
        // MapDownloadDialog.OpenMap), split in two: priming stores the
        // level selection on the solo flow coordinator (applied on its next
        // activation), pressing the Solo button only works once the main
        // menu is visible again.
        public static bool PrimeSoloFlow(BeatmapLevel level)
        {
            if (level == null) return false;

            var customLevelsPack = Loader.CustomLevelsPack;
            if (customLevelsPack == null || customLevelsPack._beatmapLevels == null || customLevelsPack._beatmapLevels.Length == 0)
                return false;

            var soloFlowCoordinator = Resources.FindObjectsOfTypeAll<SoloFreePlayFlowCoordinator>().FirstOrDefault();
            if (soloFlowCoordinator == null)
            {
                Plugin.Log?.Error("SoloFreePlayFlowCoordinator not found");
                return false;
            }

            var state = new LevelSelectionFlowCoordinator.State(
                SelectLevelCategoryViewController.LevelCategory.All, customLevelsPack, default, level);
            soloFlowCoordinator.Setup(state);
            return true;
        }

        public static bool PressSoloButton()
        {
            var songSelectButton = GameObject.Find("SoloButton")
                ?? GameObject.Find("Wrapper/BeatmapWithModifiers/BeatmapSelection/EditButton");
            if (songSelectButton == null)
            {
                Plugin.Log?.Error("Could not find the solo menu button to press");
                return false;
            }

            var button = songSelectButton.GetComponent<NoTransitionsButton>()
                ?? songSelectButton.GetComponentInChildren<NoTransitionsButton>();
            if (button == null)
            {
                Plugin.Log?.Error("Solo menu object found but has no NoTransitionsButton");
                return false;
            }

            button.onClick.Invoke();
            return true;
        }

        private static SongInstallResult Fail(string error) => new SongInstallResult { Success = false, Error = error };

        private sealed class BeatSaverMap { public List<BeatSaverVersion> versions { get; set; } }
        private sealed class BeatSaverVersion { public string hash { get; set; } public string downloadURL { get; set; } }
    }
}

File: src/SongInstaller.cpp
#include "SongInstaller.hpp"
#include "Web.hpp"
#include "main.hpp"

#include "beatsaber-hook/shared/config/rapidjson-utils.hpp"

#include "songcore/shared/SongCore.hpp"

#include "GlobalNamespace/LevelSelectionFlowCoordinator.hpp"
#include "GlobalNamespace/SelectLevelCategoryViewController.hpp"
#include "GlobalNamespace/SoloFreePlayFlowCoordinator.hpp"
#include "HMUI/NoTransitionsButton.hpp"
#include "System/Nullable_1.hpp"
#include "UnityEngine/GameObject.hpp"
#include "UnityEngine/Object.hpp"

#include "zip.h"

#include <algorithm>
#include <filesystem>
#include <thread>

using namespace GlobalNamespace;

namespace SnipeFeed::Installer {

    static constexpr long META_TIMEOUT = 15;
    static constexpr long DOWNLOAD_TIMEOUT = 120;

    static std::string ToLower(std::string s) {
        std::transform(s.begin(), s.end(), s.begin(), [](unsigned char c) { return std::tolower(c); });
        return s;
    }

    GlobalNamespace::BeatmapLevel* GetInstalledLevel(std::string const& hash) {
        if (hash.empty()) return nullptr;
        auto level = SongCore::API::Loading::GetLevelByHash(hash);
        if (!level)
            level = SongCore::API::Loading::GetLevelByHash(ToLower(hash));
        return level;
    }

    // The hash arrives from feed JSON; it becomes a URL segment and an
    // install folder name, so anything but exactly 40 hex chars is refused.
    static bool IsValidHash(std::string const& hash) {
        if (hash.size() != 40) return false;
        return std::all_of(hash.begin(), hash.end(), [](unsigned char c) {
            return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
        });
    }

    // Per-entry extraction instead of zip_stream_extract: the bulk helper
    // recreates symlink entries with unrestricted targets, letting a hostile
    // archive write through them anywhere on disk. Writing each regular file
    // ourselves (zip_entry_fread) closes that door, and lets us refuse
    // escape-y names and cap the decompressed size.
    static bool ExtractArchiveSafely(std::string const& zipData, std::string const& targetDir, std::string& outError) {
        constexpr unsigned long long MAX_TOTAL_BYTES = 512ull * 1024 * 1024;

        zip_t* archive = zip_stream_open(zipData.data(), zipData.size(), 0, 'r');
        if (!archive) {
            outError = "The map archive could not be read.";
            return false;
        }
        int total = zip_entries_total(archive);
        if (total <= 0) {
            zip_close(archive);
            outError = "The map archive is empty.";
            return false;
        }

        unsigned long long extractedBytes = 0;
        for (int i = 0; i < total; i++) {
            if (zip_entry_openbyindex(archive, i) != 0) {
                zip_close(archive);
                outError = "The map archive could not be read.";
                return false;
            }
            char const* rawName = zip_entry_name(archive);
            std::string name = rawName ? rawName : "";
            bool isDir = zip_entry_isdir(archive) == 1;

            unsigned long long entrySize = zip_entry_size(archive);
            // A forged zip64 size could wrap the running total past the cap.
            if (entrySize > MAX_TOTAL_BYTES) {
                zip_entry_close(archive);
                zip_close(archive);
                outError = "The map archive is unreasonably large.";
                return false;
            }
            extractedBytes += entrySize;

            // Stricter than strictly necessary ("song..egg" is also refused)
            // — real beatmap zips contain plain relative names only.
            bool badName = name.empty()
                || name.front() == '/'
                || name.find('\\') != std::string::npos
                || name.find("..") != std::string::npos
                || name.find(':') != std::string::npos;
            if (badName || extractedBytes > MAX_TOTAL_BYTES) {
                zip_entry_close(archive);
                zip_close(archive);
                outError = badName ? "The map archive contains an unsafe file name."
                                   : "The map archive is unreasonably large.";
                return false;
            }

            std::string destination = targetDir + "/" + name;
            std::error_code fsError;
            if (isDir) {
                std::filesystem::create_directories(destination, fsError);
            } else {
                auto parent = std::filesystem::path(destination).parent_path();
                if (!parent.empty()) std::filesystem::create_directories(parent, fsError);
                if (zip_entry_fread(archive, destination.c_str()) != 0) {
                    zip_entry_close(archive);
                    zip_close(archive);
                    outError = "Failed to extract the map archive.";
                    return false;
                }
            }
            zip_entry_close(archive);
        }
        zip_close(archive);
        return true;
    }

    void DownloadAndInstallAsync(std::string hash, std::function<void(bool, std::string)> onDone) {
        std::thread worker([hash = ToLower(std::move(hash)), onDone = std::move(onDone)] {
            // onDone must fire exactly once; if the callback itself
            // throws, the catch below must not fire it again.
            bool doneCalled = false;
            auto finish = [&](bool ok, std::string message) {
                doneCalled = true;
                onDone(ok, std::move(message));
            };

            try {

                if (!IsValidHash(hash)) {
                    finish(false, "This score has an invalid map hash.");
                    return;
                }

                // 1) Look the map up on BeatSaver.
                std::string meta;
                long code = Web::Get("https://api.beatsaver.com/maps/hash/" + hash, META_TIMEOUT, meta);
                if (code < 0) {
                    finish(false, "Network error. Check your connection.");
                    return;
                }
                if (code == 404) {
                    finish(false, "Map not found on BeatSaver.");
                    return;
                }
                if (code != 200) {
                    finish(false, "BeatSaver lookup failed (HTTP " + std::to_string(code) + ").");
                    return;
                }

                rapidjson::Document doc;
                doc.Parse(meta);
                if (doc.HasParseError() || !doc.IsObject() || !doc.HasMember("versions") || !doc["versions"].IsArray() || doc["versions"].GetArray().Empty()) {
                    finish(false, "Unexpected BeatSaver response.");
                    return;
                }

                // Only the version matching the score's hash: a different
                // version would never match the post-install lookup and
                // silently changes the map under the score.
                std::string downloadUrl;
                for (auto const& version : doc["versions"].GetArray()) {
                    if (!version.IsObject() || !version.HasMember("downloadURL") || !version["downloadURL"].IsString())
                        continue;
                    if (version.HasMember("hash") && version["hash"].IsString() && ToLower(version["hash"].GetString()) == hash) {
                        downloadUrl = version["downloadURL"].GetString();
                        break;
                    }
                }
                if (downloadUrl.empty()) {
                    finish(false, "This score's map version is no longer available on BeatSaver.");
                    return;
                }

                // 2) Download the zip (binary-safe std::string buffer).
                std::string zipData;
                code = Web::Get(downloadUrl, DOWNLOAD_TIMEOUT, zipData);
                if (code < 0) {
                    finish(false, "Network error while downloading the map.");
                    return;
                }
                if (code != 200 || zipData.empty()) {
                    finish(false, "Download failed (HTTP " + std::to_string(code) + ").");
                    return;
                }

                // 3) Extract into a temp sibling and rename into place, so a
                //    failure never leaves a half-written folder for SongCore
                //    to choke on, and a retry never merges with stale files.
                auto levelsRoot = std::string(SongCore::API::Loading::GetPreferredCustomLevelPath());
                auto targetFolder = levelsRoot + "/" + hash;
                auto tempFolder = levelsRoot + "/.snipefeed_tmp_" + hash;

                std::error_code fsError;
                std::filesystem::remove_all(tempFolder, fsError);
                std::filesystem::create_directories(tempFolder, fsError);
                if (fsError) {
                    finish(false, "Couldn't create the install folder.");
                    return;
                }

                std::string extractError;
                if (!ExtractArchiveSafely(zipData, tempFolder, extractError)) {
                    std::filesystem::remove_all(tempFolder, fsError);
                    finish(false, extractError);
                    return;
                }

                std::filesystem::remove_all(targetFolder, fsError);
                fsError.clear();
                std::filesystem::rename(tempFolder, targetFolder, fsError);
                if (fsError) {
                    std::filesystem::remove_all(tempFolder, fsError);
                    finish(false, "Couldn't move the map into Custom Levels.");
                    return;
                }

                SnipeFeedLogger.info("Installed map {} to {}", hash, targetFolder);
                finish(true, "");
            } catch (std::exception const& e) {
                // An exception escaping a detached thread is std::terminate.
                SnipeFeedLogger.error("Map install crashed: {}", e.what());
                if (!doneCalled)
                    onDone(false, "Something went wrong installing the map.");
            }
        });
        worker.detach();
    }

    // Mirrors BeatLeader's MapDownloadDialog::OpenMap for our stack, split in
    // two: priming must happen while the solo flow coordinator is still
    // active (FindObjectOfType only sees active objects), pressing the Solo
    // button only works once the main menu is visible again.
    bool PrimeSoloFlow(GlobalNamespace::BeatmapLevel* level) {
        if (level == nullptr) return false;

        auto customLevelsPack = SongCore::API::Loading::GetCustomLevelPack();
        if (customLevelsPack == nullptr) return false;
        if (customLevelsPack->_beatmapLevels->get_Length() == 0) return false;

        auto levelCategory = System::Nullable_1<SelectLevelCategoryViewController::LevelCategory>();
        levelCategory.value = SelectLevelCategoryViewController::LevelCategory(SelectLevelCategoryViewController::LevelCategory::All);
        levelCategory.hasValue = true;

        auto state = LevelSelectionFlowCoordinator::State::New_ctor(customLevelsPack, level);
        state->___levelCategory = levelCategory;

        auto soloFlowCoordinator = UnityEngine::Object::FindObjectOfType<SoloFreePlayFlowCoordinator*>();
        if (!soloFlowCoordinator) {
            SnipeFeedLogger.error("SoloFreePlayFlowCoordinator not found");
            return false;
        }
        soloFlowCoordinator->Setup(state);
        return true;
    }

    bool PressSoloButton() {
        SafePtrUnity<UnityEngine::GameObject> songSelectButton = UnityEngine::GameObject::Find("SoloButton").unsafePtr();
        if (!songSelectButton)
            songSelectButton = UnityEngine::GameObject::Find("Wrapper/BeatmapWithModifiers/BeatmapSelection/EditButton");
        if (!songSelectButton) {
            SnipeFeedLogger.error("Could not find the solo menu button to press");
            return false;
        }
        auto button = songSelectButton->GetComponent<HMUI::NoTransitionsButton*>();
        if (!button)
            button = songSelectButton->GetComponentInChildren<HMUI::NoTransitionsButton*>();
        if (!button) {
            SnipeFeedLogger.error("Solo menu object found but has no NoTransitionsButton");
            return false;
        }
        button->Press();
        return true;
    }
}
